### PR TITLE
feat(datasource/azure-pipelines-tasks): Adds support for the Azure DevOps Marketplace

### LINKED ISSUE #19694: Add support for the Azure DevOps Marketplace tasks
### What would you like Renovate to be able to do?

Currently Renovate only updates the built-in tasks supplied by Microsoft and ignores any task that's installed through the Azure DevOps Marketplace.

Renovate should also suggest updates for any public marketplace installed tasks.

### If you have any ideas on how this should be implemented, please tell us here.

* https://github.com/renovatebot/renovate/pull/19692

### Is this a feature you are interested in implementing yourself?

Yes

@JamieMagee To transfer the repo to generate the static file I can initiate a repository transfer:

 * [x] Transfer `https://github.com/jessehouwing/renovate-azure-devops-marketplace/` to `github.com/renovatebot/azure-devops-marketplace/`
 * [x] https://github.com/renovatebot/renovate/pull/19764
 * [x] Set a PAT in the Actions Secret
 * [x] Change the commit details in the repository scripts
 * [x] Add me as contributor

### PR DESCRIPTION
**Pull request by jessehouwing**
## Changes

Adds support for Azure DevOps Marketplace
Fetches marketplace data from https://github.com/jessehouwing/renovate-azure-devops-marketplace which refreshes daily

## Context

Renovate currently doesn't support the non-built-in tasks. My own security research shows how many older tasks have significant issues and conversations with admins shows it's very hard for admins to ensure the latest versions of tasks are used.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by JamieMagee** [comment]
Overall this looks good. We might just want to migrate your script into this repo directly, but I don't see this as a blocker for now.

**Review by Snehasish86 — DISMISSED** [review]

File: lib/modules/datasource/azure-pipelines-tasks/index.ts [hunk 1]
```diff
@@ -22,7 +27,10 @@ export class AzurePipelinesTasksDatasource extends Datasource {
   getReleases({
     packageName,
   }: GetReleasesConfig): Promise<ReleaseResult | null> {
-    const versions = this.builtInTasks[packageName.toLowerCase()];
+    const versions =
+      this.builtInTasks[packageName.toLowerCase()] ||
```
Comment (viceice) [inline lib/modules/datasource/azure-pipelines-tasks/index.ts:31]:
```suggestion
      this.builtInTasks[packageName.toLowerCase()] ??
```

File: tools/static-data/generate-azure-pipelines-marketplace-tasks.mjs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import got from 'got';
```
Comment (viceice) [inline tools/static-data/generate-azure-pipelines-marketplace-tasks.mjs:1]: this file needs to be called from our auto update GitHub actions workflow

Comment (JamieMagee) [reply tools/static-data/generate-azure-pipelines-marketplace-tasks.mjs:1]: Add an extra entry here:

https://github.com/renovatebot/renovate/blob/d638ed7f8898879bbb61efc029ea5eedc94dd742/package.json#L52-L55

Comment (viceice) [reply tools/static-data/generate-azure-pipelines-marketplace-tasks.mjs:1]: i think it should be combined with existing azure update script

**Review by JamieMagee — APPROVED** [review]

**Comment by jessehouwing** [comment]
> Overall this looks good. We might just want to migrate your script into this repo directly, but I don't see this as a blocker for now.

The script caches all extension an task manifests to speed things up. Which was way more data than I thought, I'd be willing to transfer this repo to the renovate org after cleaning it up a bit more though. I have a few ideas to speed up the update process (which currently takes about 10 to 15 minutes for each run).

Without the cache the update takes more than 24h as-is and transfers about 650GB of data.

The cache stores the minimum files to be able to cleanly regenerate the renovate-data file, but still adds almost 100k files to the repo and slows things down considerably.

image

My recommendation: bring the repo into the renovate org, but keep it separate from the main renovate repo. Or I need to refactor it a bit to stick the cache into a separate repo and update that from the workflow.

To run the updater, you also need to configure an Azure DevOps Marketplace PAT with `Marketplace (Read)` permissions stored as a GitHub Actions Secret.

**Comment by JamieMagee** [comment]
Wow :eyes: I didn't realise it was such a large amount of data.

That sounds like a good suggestion. We can create a followup issue to do that after this PR is merged.

**Comment by jessehouwing** [comment]
I hadn't anticipated that either when I got started :o).

**Review by viceice — APPROVED** [review]

File: tools/static-data/generate-azure-pipelines-marketplace-tasks.mjs [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+import shell from 'shelljs';
+import { updateJsonFile } from './utils.mjs';
+
+const dataUrl =
+  'https://raw.githubusercontent.com/jessehouwing/renovate-azure-devops-marketplace/main/azure-pipelines-marketplace-tasks.json';
+
+await (async () => {
+  shell.echo('Generating azure pipelines marketplace tasks');
+  const { body } = await got(dataUrl);
+  await updateJsonFile('./data/azure-pipelines-marketplace-tasks.json', body);
```
Comment (jessehouwing) [inline tools/static-data/generate-azure-pipelines-marketplace-tasks.mjs:11]: Do you want to call 

```
  const data = JSON.stringify(
    body,
    (_, value) => (value instanceof Set ? [...value] : value),
    2
  );
```

Here as well?

**Pull request merged**